src/Microsoft.ML.StandardLearners/FactorizationMachine/FactorizationMachineTrainer.cs

```diff
@@ -122,13 +131,26 @@ public FieldAwareFactorizationMachineTrainer(IHostEnvironment env, Arguments arg
         {
             Initialize(env, args);
             Info = new TrainerInfo(supportValid: true, supportIncrementalTrain: true);
+
+            // There can be multiple feature columns in FFM, jointly specified by args.FeatureColumn and args.ExtraFeatureColumns.
+            FeatureColumns = new SchemaShape.Column[1 + args.ExtraFeatureColumns.Length];
+
+            // Treat the default feature column as the 1st field.
+            FeatureColumns[0] = new SchemaShape.Column(args.FeatureColumn, SchemaShape.Column.VectorKind.Vector, NumberType.R4, false);
+
+            // Add 2nd, 3rd, and other fields from a FFM-specific argument, args.ExtraFeatureColumns.
+            for (int i = 0; args.ExtraFeatureColumns != null && i < args.ExtraFeatureColumns.Length; i++)
```

Comment:
I guess it can't. The array size is pre-defined and I prefer not to create an intermediate list just for calling `Add`.

---
In reply to: [250316802](https://github.com/dotnet/machinelearning/pull/2205#discussion_r250316802) [](ancestors = 250316802)